### PR TITLE
Bug fix for migration of disjunction clauses

### PR DESCRIPTION
**Pull request by gandhek**
*Issue #, if available:*
https://github.com/aws-cloudformation/cloudformation-guard/issues/135

*Description of changes:*

Incorrect migration of |OR| clauses from 1.0 => 2.0

Consider the template below
```
{
    "Resources": {
        "NewVolume" : {
            "Type" : "AWS::EC2::Volume",
            "Properties" : {
                "Size" : 100,
                "Encrypted": true,
                "AvailabilityZone" : "us-west-2b"
            }
        },
        "NewVolume2" : {
            "Type" : "AWS::EC2::Volume",
            "Properties" : {
                "Size" : 50,
                "Encrypted": true,
                "AvailabilityZone" : "us-west-2c"
            }
        }
    }
}
```

Guard 1.0 Rule:

```
let encryption_flag = true

AWS::EC2::Volume Encrypted == %encryption_flag
AWS::EC2::Volume Size == 100 |OR| AWS::EC2::Volume Size == 50
```

When running the rule against the example template using Guard 1.0, the result is "PASS"

Using the migrate functionality of Guard 2.0, the below rule is generated:

```
let aws_ec2_volume = Resources.*[ Type == "AWS::EC2::Volume" ]
        let encryption_flag = true

    %aws_ec2_volume.Properties.Encrypted == %encryption_flag
    %aws_ec2_volume.Properties.Size == 100 or %aws_ec2_volume.Properties.Size == 50
```
When running the migrated rule against the example template using Guard 2.0, the result is "FAIL" as the generated rule checks for ALL ec2 volume property size == 100 or ALL ec2 volume property size == 50 as opposed to ec2 property size of any == 100 or 50.

The PR fixes this for basic rules.

With this PR, the following rule is generated instead

```

let encryption_flag = true
let aws_ec2_volume = Resources.*[ Type == "AWS::EC2::Volume" ]
rule aws_ec2_volume_checks WHEN %aws_ec2_volume NOT EMPTY {
    %aws_ec2_volume {
        Properties.Encrypted == %encryption_flag
        Properties.Size == 100 or Properties.Size == 50
    }
}

```

The following PR does not fix (and acknowledges) the below gaps in the migrate command
1. The command is currently unable to add custom error messages for Conditional Clauses for rules to be migrated from 1.0 => 2.0
2. The current change will NOT work for dis junction of clauses having two different types. For e.g. a 1.0 clause for EC2 when ORed with a clause for S3 will result in an incorrectly migrated 2.0 clause.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: guard/src/migrate/parser.rs [hunk 1]
```diff
@@ -147,13 +153,52 @@ impl Display for Rule {
...
+
+        for (anchor,rules) in property_to_rules.into_iter() {
+            if rules.len() == 1 {
+                result.push(format!("{}", rules[0]))
+            } else {
+                let anchor_value = anchor.type_name;
+                let mut result_for_key= Vec::new();
+                for(rule) in rules.iter() {
+                    match rule {
+                        Rule::Conditional(conditional) => {},
```
Comment (dchakrav-github) [inline guard/src/migrate/parser.rs:193]: This should never occur based on the code above from 176 - 181. Just use the `if-let` form for these use cases, like `if let Rule::Basic(basic) = rule {...}`

File: guard/src/migrate/parser_tests.rs [hunk 2]
```diff
@@ -392,6 +393,75 @@ fn test_parse_rules_file_rule_error() {
...
+        }),
+        RuleLineType::Clause(Clause {
+            rules: vec![volume_size_100, volume_size_50]
+        })
+    ];
+
+    let actual_rule_str: Vec<String> = actual_rules.clone().into_iter().map(|rule| format!("{}", rule)).collect();
+    let migrated_rule_str = format!("{}", actual_rule_str.join("\n"));
+
+    let expected_rule_str = "\tlet encryption_flag = true\n%aws_ec2_volume.Properties.Encrypted == \"%encryption_flag\"\n%aws_ec2_volume {\n\t\tProperties.Size == 100 or Properties.Size == 50\n\t}";
```
Comment (dchakrav-github) [inline guard/src/migrate/parser_tests.rs:453]: See above comment, this must not be the form for `"%encryption_flag"`. With the quotes, it is string not variable access

**Review by dchakrav-github — CHANGES_REQUESTED** [review]

File: guard/src/migrate/parser.rs [hunk 3]
```diff
@@ -98,36 +95,45 @@ impl Display for PropertyComparison {
     }
 }
 
+#[derive(Ord, Eq, PartialEq, PartialOrd, Debug, Clone, Hash)]
```
[the remark this answers is not in the document]
Comment (gandhek) [reply guard/src/migrate/parser.rs:98]: Having a TypeName and a custom display for it seems more consistent with the other constructs than having to do this every time we print.

File: guard/src/migrate/parser.rs [hunk 4]
```diff
@@ -147,13 +153,52 @@ impl Display for Rule {
...
+
 #[derive(PartialEq, Debug, Clone, Hash)]
 pub(crate) struct Clause {
     pub(crate) rules: Vec<Rule>
 }
 impl Display for Clause {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let result: Vec<String> = self.rules.clone().into_iter().map(|rule| format!("{}", rule)).collect();
+
+        let mut property_to_rules = HashMap::new();
```
[the remark this answers is not in the document]
Comment (gandhek) [reply guard/src/migrate/parser.rs]: Agreed, incorporated changes to ensure that Display does only writing

File: guard/src/migrate/parser.rs [hunk 5]
```diff
@@ -98,37 +95,49 @@ impl Display for PropertyComparison {
...
+    pub type_name: TypeName,
     pub when_condition: PropertyComparison,
     pub check_condition: PropertyComparison
 }
 impl Display for ConditionalRule {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        writeln!(f, "when %{}[ {} ] not EMPTY {{", self.type_name.to_lowercase().replace("::", "_"), self.when_condition);
-        writeln!(f, "\t\t%{}[ {} ].{}", self.type_name.to_lowercase().replace("::", "_"), self.when_condition, self.check_condition);
-        write!(f, "\t}}")
+        writeln!(f, "when %{}[ {} ] not EMPTY {{", self.type_name, self.when_condition);
```
Comment (dchakrav-github) [inline guard/src/migrate/parser.rs:135]: If i understand the change right, you already have an external `%aws_ec2_volume { ... }` block. Inside the block `this` already refers each `AWS::EC2::Volume`, does this `when` check need the filter part? It seems like we will generate

```
      %aws_ec2_volume {
            %aws_ec2_volume[ Properties.Size > 500 ] NOT EMPTY {
                  %aws_ec2_volume[ Size > 500 ].Encrypted == false
            }
      }
```

We should just generate 
```
      %aws_ec2_volume {
             when Properties.Size > 500 {
                     Properties.Encrypted == false
             }
      }
```
Easier to comprehend as well.

File: guard/src/commands/migrate_tests.rs [hunk 6]
```diff
@@ -86,4 +89,56 @@ fn test_migrate_rules() -> Result<()> {
     let span = crate::rules::parser::Span::new_extra(&result, "");
     rules_file(span)?;
     Ok(())
+}
```
Comment (dchakrav-github) [inline guard/src/commands/migrate_tests.rs:92]: Please add tests for Conditional Rules (CR), rule with `OR`s of CR, rule with `OR`s with Conditional Rule and a Basic rule combination.

File: guard/src/migrate/parser.rs [hunk 7]
```diff
@@ -180,10 +192,9 @@ impl Display for RuleLineType {
 
 // variable-dereference =  ("%" variable) / ("%{" variable "}" ); Regular and environment variables, respectively
 pub (crate) fn parse_variable_dereference(input: Span) -> IResult<Span, String> {
-    preceded(space0, alt((
+    delimited(space0, alt((
```
Comment (dchakrav-github) [inline guard/src/migrate/parser.rs:195]: Please add an explanation on why this change was needed for the next set of maintainers. It is hard to understand why this change was made without the context surrounding this

File: guard/src/migrate/parser_tests.rs [hunk 8]
```diff
@@ -392,6 +393,65 @@ fn test_parse_rules_file_rule_error() {
     );
 }
 
+#[test]
+fn test_disjunction_basic_clauses() {
```
Comment (dchakrav-github) [inline guard/src/migrate/parser_tests.rs:397]: Please add tests for Conditional Rules and disjunction combinations across them as well.

**Review by dchakrav-github — CHANGES_REQUESTED** [review]

File: guard/src/commands/migrate_tests.rs [hunk 9]
```diff
@@ -141,4 +176,48 @@ rule aws_s3_bucket_checks WHEN %aws_s3_bucket NOT EMPTY {
...
+    let expected_rule = String::from("let aws_ec2_instance = Resources.*[ Type == \"AWS::EC2::Instance\" ]
+rule aws_ec2_instance_checks WHEN %aws_ec2_instance NOT EMPTY {
+    %aws_ec2_instance {
+        when Properties.InstanceType == \"m2.large\" {
+            DeletionPolicy == \"Retain\"
+        } or Properties.InstanceType == \"t2.micro\" <<Deletion policy for m2.large>>
+    }
+}
+\n");
+    assert_eq!(result, expected_rule);
```
Comment (dchakrav-github) [inline guard/src/commands/migrate_tests.rs:220]: Comparing strings makes it very difficult to maintain. Create a `RulesFile` object from both migrated and expected strings and compares these objects

File: guard/src/commands/migrate_tests.rs [hunk 10]
```diff
@@ -114,6 +122,33 @@ rule aws_ec2_volume_checks WHEN %aws_ec2_volume NOT EMPTY {
...
+        } or when Properties.InstanceType == "t2.micro" {
+            Encrypted == true
+        }
+    }
+}
+
+"#);
+    println!("{}", result);
+    assert_eq!(result, expected_rule);
+    rules_file(span)?;
```
Comment (dchakrav-github) [inline guard/src/commands/migrate_tests.rs:148]: same comment as below parse both strings and compare the parsed objects

File: guard/src/migrate/parser.rs (hunk 7 above)
Comment (dchakrav-github) [reply guard/src/migrate/parser.rs:193]: I get that gist from the change made in the code :). Elaborate on the context for 
1. Why was stripping white spaces needed? 
2. What was the error/consequence when this wasn't done?

**Review by dchakrav-github — CHANGES_REQUESTED** [review]

**Review by dchakrav-github — APPROVED** [review]

**Review by priyap286 — APPROVED** [review]

**Pull request merged**